File: web/src/api/guru.ts
import { apiClient } from './client';

/**
 * Guru Mode API Service
 * 
 * Handles all Guru Mode (Teach AI) related API calls
 * for the Feynman Technique teaching feature.
 */

// =============================================================================
// Types
// =============================================================================

export type GuruPersona = '5-year-old' | 'peer' | 'skeptic' | 'curious_beginner';

export interface GuruSessionCreate {
  topic: string;
  subject: string;
  persona?: GuruPersona;
}

export interface GuruSessionResponse {
  session_id: string;
  topic: string;
  subject: string;
  persona: string;
  initial_message: string;
  created_at: string;
}

export interface ChatMessage {
  role: 'user' | 'model';
  content: string;
}

export interface GuruChatRequest {
  session_id: string;
  message: string;
}

export interface GuruChatResponse {
  message: string;
  confusion_level: number; // 0-100
  is_satisfied: boolean;
  hints?: string[] | null;
}

export interface GuruReportCard {
  accuracy_score: number; // 0-10
  simplicity_score: number; // 0-10
  feedback: string;
  xp_earned: number;
  strengths?: string[] | null;
  improvements?: string[] | null;
}

export interface GuruEndSessionResponse {
  session_id: string;
  status: string;
  report_card: GuruReportCard;
  total_messages: number;
  session_duration_seconds: number;
}

export interface GuruSessionSummary {
  session_id: string;
  subject: string;
  topic: string;
  persona: string;
  status: 'active' | 'completed' | 'abandoned';
  accuracy_score?: number | null;
  simplicity_score?: number | null;
  xp_earned: number;
  created_at: string;
  message_count: number;
}

export interface GuruHistoryResponse {
  sessions: GuruSessionSummary[];
  total_sessions: number;
  total_xp_earned: number;
  average_accuracy?: number | null;
  average_simplicity?: number | null;
}

export interface GuruSessionDetailResponse {
  session_id: string;
  subject: string;
  topic: string;
  persona: string;
  status: string;
  messages: ChatMessage[];
  report_card?: GuruReportCard | null;
  xp_earned: number;
  created_at: string;
  updated_at: string;
}

export interface ActiveSessionResponse {
  has_active: boolean;
  session_id?: string;
  topic?: string;
  subject?: string;
  persona?: string;
  created_at?: string;
}

// =============================================================================
// API Functions
// =============================================================================

/**
 * Start a new Guru Mode teaching session
 */
export async function startGuruSession(request: GuruSessionCreate): Promise<GuruSessionResponse> {
  const response = await apiClient.post<GuruSessionResponse>('/guru/start', request);
  return response.data;
}

/**
 * Send a teaching message in an active session
 */
export async function sendGuruMessage(request: GuruChatRequest): Promise<GuruChatResponse> {
  const response = await apiClient.post<GuruChatResponse>('/guru/chat', request);
  return response.data;
}

/**
 * End a session and get the report card
 */
export async function endGuruSession(sessionId: string): Promise<GuruEndSessionResponse> {
  const response = await apiClient.post<GuruEndSessionResponse>('/guru/end', {
    session_id: sessionId,
  });
  return response.data;
}

/**
 * Abandon a session without grading
 */
export async function abandonGuruSession(sessionId: string): Promise<{ message: string }> {
  const response = await apiClient.post<{ message: string }>('/guru/abandon', {
    session_id: sessionId,
  });
  return response.data;
}

/**
 * Get user's Guru session history
 */
export async function getGuruHistory(
  limit: number = 20,
  offset: number = 0
): Promise<GuruHistoryResponse> {
  const response = await apiClient.get<GuruHistoryResponse>('/guru/history', {
    params: { limit, offset },
  });
  return response.data;
}

/**
 * Get details of a specific session
 */
export async function getGuruSessionDetail(sessionId: string): Promise<GuruSessionDetailResponse> {
  const response = await apiClient.get<GuruSessionDetailResponse>(`/guru/session/${sessionId}`);
  return response.data;
}

/**
 * Check if user has an active session
 */
export async function getActiveGuruSession(): Promise<ActiveSessionResponse> {
  const response = await apiClient.get<ActiveSessionResponse>('/guru/active');
  return response.data;
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Get display name for persona
 */
export function getPersonaDisplayName(persona: GuruPersona | string): string {
  const names: Record<string, string> = {
    '5-year-old': '5-Year-Old Child',
    'peer': 'Fellow Student',
    'skeptic': 'Skeptical Student',
    'curious_beginner': 'Curious Beginner',
  };
  return names[persona] || persona;
}

/**
 * Get persona description
 */
export function getPersonaDescription(persona: GuruPersona | string): string {
  const descriptions: Record<string, string> = {
    '5-year-old': 'Needs very simple explanations with no big words',
    'peer': 'A classmate who missed the lesson and needs catching up',
    'skeptic': 'Questions everything and asks for proof',
    'curious_beginner': 'Excited to learn and asks thoughtful follow-ups',
  };
  return descriptions[persona] || '';
}

/**
 * Get color based on confusion level
 */
export function getConfusionColor(level: number): string {
  if (level <= 20) return 'text-green-400';
  if (level <= 40) return 'text-emerald-400';
  if (level <= 60) return 'text-yellow-400';
  if (level <= 80) return 'text-orange-400';
  return 'text-red-400';
}

/**
 * Get confusion status text
 */
export function getConfusionStatus(level: number): string {
  if (level <= 20) return 'Understands well!';
  if (level <= 40) return 'Getting there...';
  if (level <= 60) return 'Still confused';
  if (level <= 80) return 'Very confused';
  return 'Totally lost!';
}

File: backend/app/schemas/guru.py
"""
Pydantic schemas for Guru Mode (Teach AI)

Handles request/response validation for the teaching session feature
where students teach concepts to an AI persona using the Feynman Technique.
"""
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from enum import Enum
from uuid import UUID


class GuruPersona(str, Enum):
    """Available AI student personas"""
    FIVE_YEAR_OLD = "5-year-old"
    PEER = "peer"
    SKEPTIC = "skeptic"
    CURIOUS_BEGINNER = "curious_beginner"


class GuruSessionStatus(str, Enum):
    """Session status options"""
    ACTIVE = "active"
    COMPLETED = "completed"
    ABANDONED = "abandoned"


# ============================================================================
# Session Management
# ============================================================================

class GuruSessionCreate(BaseModel):
    """Request to start a new Guru Mode teaching session"""
    topic: str = Field(..., min_length=1, max_length=100, description="Topic to teach")
    subject: str = Field(..., min_length=1, max_length=50, description="Subject area")
    persona: Optional[GuruPersona] = Field(
        default=GuruPersona.PEER,
        description="AI student persona type"
    )


class GuruSessionResponse(BaseModel):
    """Response after starting a Guru session"""
    session_id: str
    topic: str
    subject: str
    persona: str
    initial_message: str
    created_at: datetime


# ============================================================================
# Chat Messages
# ============================================================================

class ChatMessage(BaseModel):
    """Single chat message in a session"""
    role: str = Field(..., description="'user' or 'model'")
    content: str


class GuruChatRequest(BaseModel):
    """Request to send a message in a Guru session"""
    session_id: str = Field(..., description="UUID of the active session")
    message: str = Field(..., min_length=1, max_length=2000, description="User's teaching message")


class GuruChatResponse(BaseModel):
    """Response from the AI student"""
    message: str = Field(..., description="AI's response message")
    confusion_level: int = Field(
        ..., 
        ge=0, 
        le=100, 
        description="AI's confusion level (100 = very confused, 0 = fully understood)"
    )
    is_satisfied: bool = Field(
        default=False, 
        description="Whether the AI is satisfied with the explanation"
    )
    hints: Optional[List[str]] = Field(
        default=None,
        description="Optional hints for what the AI is confused about"
    )


# ============================================================================
# Session Ending & Report Card
# ============================================================================

class GuruEndSessionRequest(BaseModel):
    """Request to end a Guru session and get the report card"""
    session_id: str = Field(..., description="UUID of the session to end")


class GuruReportCard(BaseModel):
    """Final evaluation report card for a teaching session"""
    accuracy_score: int = Field(..., ge=0, le=10, description="Accuracy of the explanation (0-10)")
    simplicity_score: int = Field(..., ge=0, le=10, description="Simplicity/clarity of the explanation (0-10)")
    feedback: str = Field(..., description="Constructive feedback on the teaching")
    xp_earned: int = Field(..., ge=0, description="XP points earned from this session")
    strengths: Optional[List[str]] = Field(default=None, description="What the student did well")
    improvements: Optional[List[str]] = Field(default=None, description="Areas for improvement")


class GuruEndSessionResponse(BaseModel):
    """Response after ending a Guru session"""
    session_id: str
    status: str
    report_card: GuruReportCard
    total_messages: int
    session_duration_seconds: int


# ============================================================================
# Session History
# ============================================================================

class GuruSessionSummary(BaseModel):
    """Summary of a past Guru session for history listing"""
    session_id: str
    subject: str
    topic: str
    persona: str
    status: str
    accuracy_score: Optional[int] = None
    simplicity_score: Optional[int] = None
    xp_earned: int = 0
    created_at: datetime
    message_count: int = 0


class GuruHistoryResponse(BaseModel):
    """Response with user's Guru session history"""
    sessions: List[GuruSessionSummary]
    total_sessions: int
    total_xp_earned: int
    average_accuracy: Optional[float] = None
    average_simplicity: Optional[float] = None


class GuruSessionDetailResponse(BaseModel):
    """Detailed view of a single Guru session"""
    session_id: str
    subject: str
    topic: str
    persona: str
    status: str
    messages: List[ChatMessage]
    report_card: Optional[GuruReportCard] = None
    xp_earned: int = 0
    created_at: datetime
    updated_at: datetime
